nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,310 @@
+    // to mask_pose point in mask_frame_
+    geometry_msgs::msg::TransformStamped transform;
+    geometry_msgs::msg::PointStamped in, out;
+    in.header.stamp = clock_->now();
+    in.header.frame_id = global_frame_;
+    in.point.x = pose.x;
+    in.point.y = pose.y;
+    in.point.z = 0;
+
+    try {
+      tf_->transform(in, out, mask_frame_, transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        logger_,
+        "BinaryFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return false;
+    }
+    mask_pose.x = out.point.x;
+    mask_pose.y = out.point.y;
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool BinaryFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
+{
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+inline int8_t BinaryFilter::getMaskData(
+  const unsigned int mx, const unsigned int my) const
+{
+  return filter_mask_->data[my * filter_mask_->info.width + mx];
+}
+
+void BinaryFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "BinaryFilter: Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(pose, mask_pose)) {
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    // Robot went out of mask range. Set "false" state by-default
+    changeState(false);
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed
+  int8_t mask_data = getMaskData(mask_robot_i, mask_robot_j);
+  // If filter_mask data is not empty, change binary state (if necessary)
+  if (
```

Comment:
Shouldn't there be some checking of the costmap filter info about how to interpret these values?

Reply:
I am not sure about the intention of this:
* If you are talking about the mask thresholds for "on" and "off", this already exists in input mask YAML-files: `trinary` mode could be used, and necessary threshold is being set for input map.
* Someone wants to set some specific value for flipping only. In this case, binary filter is easy thing, which flips some state to `true` or `false`, and I do not believe that we need to complicate the architecture (add new fields to the `nav2_msgs/msg/CostmapFilterInfo.msg` messages only because of this use-case). In my intention, the binary filter should work simply, as Keepout Filter: for "non-0" values it will flip the filter to `true`, otherwise - `false`.
* If someone want to set e.g. `UNKNOWN` values to turn-on the filter, I am not sure that this is real use-case. Where could we apply it? Even if so, the map mask could be easily changed by moving `occupied_thresh` in YAML-file to have former unknowns to be recognized by map server as occupied values.

Reply:
I meant that we have this whole standard about how to intepret the costmap 0-255 data to represent different things like keepout/speed/etc data. It seems odd that this would not use that information from the costmap filter info given to it. The unknown/free space special numbers are for cost values, but that's not what the filter mask represents by the documentation we've generated.

There should be some threshold value for on/off and using the costmap info base/multiplier to get it. By default, it could literally just be any non-0 or non-255 value so that it, in effect, works that way. But it should be setable by users.

Reply:
Okay, looks like I've understood your intention. Added arbitrary filter space values and `flip_threshold` parameter to BinaryFilter: https://github.com/ros-planning/navigation2/pull/3228/commits/8f58d33e3640e8a239a324c8baab1b8c358eb399